File: 670-maximum-swap/670.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-


class Solution:
    def maximumSwap(self, num: int) -> int:
        maxnum, maxloc, maxchange = -1, -1, 0
        for i, n in enumerate(str(num)[::-1]):
            tmp = int(n)
            if maxnum < tmp:
                maxnum = tmp
                maxloc = i
            else:
                change = (maxnum - tmp) * 10**i + (tmp - maxnum) * 10**maxloc
                maxchange = max(maxchange, change)

        return num + maxchange


if __name__ == '__main__':
    sol = Solution()
    print(sol.maximumSwap(2736))
    print(sol.maximumSwap(9973))
    print(sol.maximumSwap(9278))

File: shoope/670-maximum-swap/670.cpp
#include <iostream>
#include <cmath>
#include <climits>

using namespace std;

class Solution {
public:
    int maximumSwap(int num) {
        int maxnum = INT_MIN;
        int maxloc = -1;
        int maxchange = 0;
        int index = 0;
        int res = num;
        while (num) {
            if (maxnum < num%10) {
                maxnum = num % 10;
                maxloc = index;
            } else {
                int change = (maxnum - num%10) * pow(10, index) + (num%10 - maxnum) * pow(10, maxloc);
                maxchange = max(change, maxchange);
            }
            ++index;
            num /= 10;
        }
        return res + maxchange;
    }
};

int main(int argc, char** args) {
    Solution sol = Solution();
    cout << sol.maximumSwap(2736) << endl;
    cout << sol.maximumSwap(9973) << endl;
    cout << sol.maximumSwap(9378) << endl;
    
    return 0;
}